prov/net,uring: cancel outstanding SQEs + various bugfixes

sydidelot opened the pull request:
The PR covers the following changes:
- it implements the cancelation of outstanding SQEs when the EP is flushed.
- it fixes a bunch of bugs I found while testing the io_uring support.

With this PR, fabtests reports the same percentage of passed tests whether io_uring is enabled or not (command line `./runfabtests.sh net`).

There is still one test that randomly fails with io_uring, I'm not sure why:
```
$ FI_NET_IO_URING=1 fi_cq_data -e rdm -p "net"   -s 127.0.0.1
Waiting for CQ data from client
remote_cq_data: success
fi_cq_data: prov/net/src/xnet_progress.c:1490: xnet_close_progress: Assertion `slist_empty(&progress->event_list)' failed.
Aborted (core dumped)
```
It looks like the process receives the FI_SHUTDOWN event in the EQ but the application doesn't consume it.

shefty commented:
@sydidelot - rdm ep's manage the underlying connections internally.  So any FI_SHUTDOWN event is consumed by the rdm ep, which will use that to close the msg ep.

shefty commented on prov/net/src/xnet_progress.c:
```diff
@@ -950,17 +950,22 @@ static void xnet_progress_cqe(struct xnet_progress *progress,
 			if (ep->cur_rx.entry)
 				xnet_complete_rx(ep, cqe->res);
-			else
+			else {
```
nit: The coding standard is to add braces to both the if and else statements if either needs them.

Maybe we should use a goto disable in this call to jump to a common point.

sydidelot replied:
Thanks. I updated the code to use a goto, as you suggested.

shefty commented on prov/net/src/xnet_progress.c:
```diff
@@ -1345,6 +1346,8 @@ static int xnet_init_uring(struct xnet_uring *uring, size_t entries,
 
 	uring->fid.fclass = XNET_CLASS_URING;
+	uring->sockapi = sockapi;
+	uring->sockapi->credits = ofi_uring_sq_space_left(&uring->ring);
```
This change is in the wrong patch, or should be in a new patch, or needs some description in the commit message regarding why it's needed.

sydidelot replied:
Thanks! I isolated this change in a new patch.

shefty commented on prov/net/src/xnet_progress.c:
```diff
@@ -1004,6 +1006,33 @@ static void xnet_progress_uring(struct xnet_progress *progress,
+				ofi_uring_submit(&uring->ring);
+				submitted = true;
+			}
+			else if (ret != -FI_EAGAIN)
```
nit: move else to same line as } and add braces to else-if statement

sydidelot replied:
Fixed. Thanks!

shefty reviewed (commented):
minor quibbles -- looks good -- thanks!

sydidelot commented:
@shefty Thanks for the review! I have updated the PR. 
In regards to the assertion failure in fi_cq_data, I think it's a side-effect of asynchronous operations with io_uring. When the connection is closed, the provider needs to poll the io_uring to find out that the connection is effectively shut down and generate the FI_SHUTDOWN event. The problem is that the generation of the FI_SHUTDOWN event may happen after the endpoint gets closed.
I added a patch to the PR where `xnet_close_progress()` destroys all events listed in `progress->event_list`. Let me know if you have a better suggestion.

shefty commented on prov/net/src/xnet_progress.c:
```diff
@@ -1479,10 +1479,17 @@ int xnet_init_progress(struct xnet_progress *progress, struct fi_info *info)
+		item = slist_remove_head(&progress->event_list);
+		event = container_of(item, struct xnet_event, list_entry);
+		free(event);
+	}
```
It might work to clean up the list for this test, but I'm concerned about the problem of having events that reference destroyed endpoints.  That could show up for other apps that have other active endpoints and don't close the progress engine.  They would read these events, try to process them, and likely access freed memory.

Maybe we need some sort of cleanup in the rdm ep close path for this instead.

sydidelot replied:
Thanks! I have removed the last commit, as you suggested. 
I will address that particular issue in a new PR.

shefty reviewed (commented):
I only see a problem with the last patch.  Given the commit message, it sounds like we can remove that and merge the other patches, since io_uring must be explicitly enabled.  We'll likely need to identify the reason the events aren't being cleaned up to see if the impact is only io_uring, or more general, with io_uring exposing some race easier.

shefty commented:
The CI failures are related to GPU memory allocation, I think.  The GPU nodes might be down.  Your last update prior to popping of the topmost patch passed CI.
